File: javascript/Strings/validParentheses.js
const isValid = (str) => {
  const stack = [];
  const relation = {
    ')': '(',
    ']': '[',
    '}': '{',
  };
  let output = false;

  str.split('').forEach((char) => {
    const last = stack[stack.length - 1];
    if (Object.keys(relation).includes(char)) {
      // Closing
      (last === relation[char]) ? stack.pop() : stack.push(char);
    } else {
      stack.push(char);
    }
  });

  if (stack.length === 0) output = true;

  return output;
};

File: ruby/valid_parentheses.rb
MAPPING = {
  ')' => '(',
  ']' => '[',
  '}' => '{'
}.tap { |h| h.default = '' }

def is_valid(string)
  stack = []
  string.each_char do |char|
    next stack.pop if MAPPING[char] == stack.last

    stack.push(char)
  end

  stack.empty?
end


p is_valid('()')
p is_valid('()[]{}')
p is_valid('(]')
p is_valid('[')